### ISSUE TITLE
Add source code to documentation

### ISSUE THREAD

**Issue by timhoffm** [matplotlib/matplotlib#10676]
# Feature Request

I propose to use the Sphinx extension [sphinx.ext.viewcode](http://www.sphinx-doc.org/en/stable/ext/viewcode.html) to add the python source code to the docs and crosslink it.

This could help users understand the relations between the various levels of indirection we use. Also it could be a little help with poorly or undocumented parts of the API (we should improve the docs, but that's a long-term project).

Here is an example how this would look like (note the `[source]` link): 

![grafik](https://user-images.githubusercontent.com/2836374/36950835-2ee81aaa-1ffc-11e8-8767-4ae6f0a782c5.png)

The link will guide to a page containing the source code:

![grafik](https://user-images.githubusercontent.com/2836374/36950848-609a0f4a-1ffc-11e8-8fd4-494c3001d5d4.png)

Please give a short feedback on the idea. If there is a general approval, I'll submit a pull request.

**Comment by jklymak** [comment]
This seems like a 👍  feature to me.  Does it take folks straight to the GitHub version or do we have to render and colorize all the source code?

**Comment by timhoffm** [comment]
sphinx.ext.viewcode does all the work of generating the code docs as highlighted html. It even adds a `[docs]` link, that will take you back to the rendered html docs.

This also ensures that docs and code are on the same version.

There's literally not more to do than adding the extension.

**Comment by jklymak** [comment]
... except that building the docs takes a long time already (for me anyways).

**Comment by timhoffm** [comment]
That's true, however in my feeling the extension doesn't make it significantly longer (I haven't timed it exactly). The long built times are another issue.

**Comment by efiring** [comment]
I'm a big fan of that extension; we should use it.  I suspect the long build time largely results from the examples--generating the plots.  I haven't looked into it, though.

**Comment by jklymak** [comment]
Yep, if it doesn't add (substantively) to build time, then this is great.  

I agree that the build time is likely a sphinx-gallery issue.

**Comment by ImportanceOfBeingErnest** [comment]
I'm totally in favor of this. 
Just a question though: Will this link to the current master code or to the code of the version which the docs refer to? E.g. if looking at the 2.0 docs, will it still link to the master, which is some versions ahead?

**Comment by timhoffm** [comment]
A copy of the code is included when building the docs. This also ensures that the html and the linked code refer to the same version. There are *no* external links (to github or any other place).

**Comment by ImportanceOfBeingErnest** [comment]
Ok I wasn't sure, since e.g. in the [numpy](docs.scipy.org/doc/numpy/reference/generated/numpy.core.defchararray.equal.html), [pandas](pandas.pydata.org/pandas-docs/stable/generated/pandas.read_json.html) and [scikit-learn](http://scikit-learn.org/stable/modules/generated/sklearn.cluster.AgglomerativeClustering.html) docs the source link directly links to the github repository while e.g. in the [astropy docs](docs.astropy.org/en/stable/api/astropy.time.TimeJD.html) it links to a html page.

**Comment by efiring** [comment]
scikit-learn does that with the 'linkcode' (not 'viewcode') extension, and with this at the bottom of the sphinx conf.py:
```
# The following is used by sphinx.ext.linkcode to provide links to github
linkcode_resolve = make_linkcode_resolve('sklearn',
                                         u'https://github.com/scikit-learn/'
                                         'scikit-learn/blob/{revision}/'
'{package}/{path}#L{lineno}')
```

**Comment by timhoffm** [comment]
Is there any benefit to linking to a fixed external resource for the code - other than saving a bit of time for generation? I have the feeling that having the code blob and the docs in one place is safer.

**Comment by jklymak** [comment]
If its not too expensive in terms of compile time, I think linking static is better.  Do the static links link to github at some point?

**Comment by efiring** [comment]
The static links do not link to github at all.  An advantage of keeping everything static is that it doesn't rely on having an internet connection.  A disadvantage is that it makes the docs bulkier.  They store the html representation of the source code, which is extremely verbose.  For example, in a small project I work with, a 17k source file yields a 91k html file.

**Comment by timhoffm** [comment]
Did some performance tests:
- Documentation build time is only slightly larger (few seconds). Actual values can vary from run to run.
- The docs grow by 10% (14MB) in size.

IMO this is bearable, so that we can just enable the extension.

### Current master
``time make html``
real    2m24,491s
user    2m12,168s
sys     0m2,760s

``du -h -d1 build/html``
28K     build/html/thirdpartypackages
3,9M    build/html/_static
260K    build/html/faq
1,1M    build/html/devel
10M     build/html/_sources
84K     build/html/mpl_toolkits
14M     build/html/gallery
1,7M    build/html/tutorials
16K     build/html/resources
24M     build/html/api
20K     build/html/glossary
4,5M    build/html/users
9,9M    build/html/_downloads
68M     build/html/_images
138M    build/html

### Current master with viewcode
``time make html``
real    2m26,873s
user    2m16,760s
sys     0m2,400s

``du -h -d1 build/html``
14M     build/html/_modules
28K     build/html/thirdpartypackages
3,9M    build/html/_static
260K    build/html/faq
1,1M    build/html/devel
10M     build/html/_sources
84K     build/html/mpl_toolkits
14M     build/html/gallery
1,7M    build/html/tutorials
16K     build/html/resources
24M     build/html/api
20K     build/html/glossary
4,5M    build/html/users
9,9M    build/html/_downloads
68M     build/html/_images
152M    build/html
